text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,409 @@
+
+In the same vein, `cargo metadata` fills the target directory information with the absolute path and make no mention of `CARGO_TARGET_BASE_DIR` since it can only be used with a single workspace at once.
+
+### `cargo clean`
+
+Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for `CARGO_TARGET_BASE_DIR`: if it set, `cargo clean` deletes `CARGO_TARGET_BASE_DIR/<project-path>/` and not `target/`.
+
+### Providing forward links
+
+[`targo`][tg] provides forward link (it links from `<workspace>/target` to its own target directory) as a way for existing tools to continue working despite there being no `CARGO_TARGET_DIR` set for them to find the real target directory.
+
+`cargo` does not provide them for `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).
+
+When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a forward link by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.
+
+This has a two big advantages: not breaking external tools and giving easy access to artifacts produced by `cargo build/test/doc` to users (they're in the habit of typing `./target/debug/my-bin`, this would continue working with forward links).
+
+A config option (CLI, `config.toml` and env var), `link-target-dir`, will be introduced to deactivate this behaviour, but it will `true` by default, for the reasons provided in favor of forward links just above.
+
+#### Detailed working of forward links
+
+When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.
```

Comment:
Wanted to call this out for reviewers that the proposal currently suggests best-effort forward-links

Reply:
For myself, I'm a little uncomfortable with the best-effort approach.  It feels sloppy.

Maybe if `link-target-dir` values were `auto` and `off`?

Reply:
I assume 'auto' would still be 'best effort' ?

Reply:
Or would it fail the command if the link cannot be created ?

Reply:
The values for `link-target-dir` are currently unspecified in the RFC.  I am making the assumption that they are `true` and `false`.  My concern is with `true` being sloppy.  I am suggesting that we have an `auto` value that is a way out to explicitly recognize that it is "sloppy".

So with the framing of having an `auto` I could see us using
- `true`, `false`, `"auto"` (bool with a special string value)
- `always`, `auto`, `never` (like colors)

If its `true`  or `always`, it would error.